### PR TITLE
implement `@expect`

### LINKED ISSUE #489: @expect() hint to optimizer
This is proposal for a small feature, local and isolated. Could improve code readability.

------------
Linux kernel uses macros `likely` and `unlikely` :

```
if (likely(x > 0)) { ... }
...
if (unlikely(y == NULL)) { ... }

```

These macros may improve performance a little bit and they serve as handy documentation.

Language Nim supports this functionality too, through its standard library ( https://nim-lang.org/docs/system.html#likely.t,bool ).

----------------------

**My proposal:**

Add `if+` and `if-` into the language. `if+` would signal that the path is rather likely, `if-` will suggest the flow will not go this way.

```
if+ (x < 0) { 
  ...  // likely goes here
}
if- (err) { 
  ... // unlikely
}

```

What is this good for:
1. It gives hint to source code reader, hint which will be almost never found in the documentation. This hint is very short, intuitive, and doesn't require extra pair of parenthesis.
2. It may slightly improve the performance, by rearranging instructions so that the likely path goes w/o jump, or by inserting hint for branch selector. 

How it could be implemented:
1. By doing nothing, just using it as documentation only feature.
2. By using IR opcode `llvm.expect`. This is how clang implements `__builtin_expect`, which is called by `likely`/`unlikely `macros. (GCC also provides `__builtin_expect`, MSVC has nothing similar.)

-------------------

Performance effects of `__builtin_expect` are hotly disputed on the internet. Many people claim programmers are invariably bad in such prediction and profile guided optimization will do much, much better. However, they never support their claim by benchmarks.

Even if performance is unaffected the documentation value remains. When one is stepping the code through debugger and the flow goes against the hint, one may be get more cautious a catch a bug.

### PR DESCRIPTION
**Pull request by Rexicon226**
closes #489 

---

#### Why it takes a bool
The `@expect` built-in affects the generation of branching control flow and you can only branch on a bool. To branch on any other value requires an existing comparison. The built-in expect from C, `built_expect`, compares two values producing a bool to the same effect as `@expect(x == y, true)`.

#### Why it doesn't have a probability
Probabilities/weights serve to re-arrange multiple branches such as a switch statement. However, there's no way to predict branching codegen in a switch control flow use case. The most likely branch could be a fallthrough, but it could also be a jump table. Only single target branches guarantee predictable branching (`if`, `while`, etc.).

---

### Testing

The `@expect` built-in should always be transitive and have no semantic effect on the condition. The tests added for the built-in include these transitive properties. A compile-error test ensures the required result type of the built-in is always a `bool`.

---

### What does it do?

This built-in is useful in code where an unwanted branch can be bad for performance. A useful example is hashing, memcpy, and similarly i-cache-sensitive applications.

The `if`'s `true` branch usually falls through, but it's not guaranteed. The `@expect` built-in can be used to guarantee this.

Here is a comparison of the function `foo`:
```zig
export fn foo(arg: *u32) u32 {
    if (@expect(arg.* == 10, true)) {
        arg.* = 10;
        return 0;
    } else {
        return arg.*;
    }
}
```
where one has `@expect` set to true, and the other has it set to false.

`true`:
```asm
; 0000000000000000 <foo>:
;     if (@expect(arg.* == 10, true)) {
       movl    (%rdi), %eax
       cmpl    $0xa, %eax
       jne    0xf <foo+0xf>
       xorl    %eax, %eax
;         arg.* = 10;
       movl    $0xa, (%rdi)
       retq
```

`false`:
```asm
; 0000000000000000 <foo>:
;     if (@expect(arg.* == 10, false)) {
       movl    (%rdi), %eax
       cmpl    $0xa, %eax
       je    0x8 <foo+0x8>
       retq
       xorl    %eax, %eax
;         arg.* = 10;
       movl    $0xa, (%rdi)
       retq
```

you can see that LLVM has generated the branch such that it falls through the jump when it's `true` or `false` respectively.

---

### Real-world use-case

A real-world example of this built-in is in implementing a better `@memcpy`. The current compiler-rt implementation isn't as fast as it could be and produces varying results. Facebook has an open-source library called "Folly," and inside there is a very fast and clean memcpy that also doubles as a memmove. They have done the tough part of figuring out an optimal branching tree, however, it is currently very hard to guarantee this in Zig. Given that the order of the branches is crucial to the performance of the implementation, we need the `@expect` built-in to influence the codegen.

Here I've implemented Folly's memcpy in Zig using the `@expect` builtin:
https://gist.github.com/Rexicon226/80048f4ccd68d0e5fce927bbbad44226

A couple of unique things to note about the usages of `@expect` here.

First, you might notice that almost every `@expect` usage is false. This is because we try to write if statements as if they are branches, not as if they are fallthroughs. It's easier to see the code flow that way and makes it easier to understand where potential branching costs could happen.

Second, there is a special part of the code that we still cannot fully emulate. This branch from the original implementation:
https://github.com/facebook/folly/blob/2d9c0a3c720238ae2ef1337348b6a72778024961/folly/memcpy.S#L131-L178

is almost possible to implement. Without the `@expect` built-in this might be possible using switch continues (#8220), however I'm not sure.

```asm
; check for len > 256, expected to be false so the case jumps away
cmp     rdx, 0x100
ja      <sinon_memcpy+0x10a>

; first 2 32-byte chunks are loaded
vmovups ymm2, ymmword ptr [rsi]
vmovups ymm0, ymmword ptr [rsi + rdx - 0x20]

; we expect the length being <= $64 here to be false, so the case jumps away
cmp     rdx, 0x41
jb      .FINAL_64

; the pattern continues here, all nicely generated from an inline for loop
vmovups ymm3, ymmword ptr [rsi + 0x20]
vmovups ymm1, ymmword ptr [rsi + rdx - 0x40]
cmp     rdx, 0x80
jbe     .FINAL_128

vmovups ymm5, ymmword ptr [rsi + 0x40]
vmovups ymm4, ymmword ptr [rsi + rdx - 0x60]
cmp     rdx, 0xc0
jbe     .FINAL_192
; ...

; the $64 case jumps to here, and finalizes the loads with these stores and returns
.FINAL_64:
vmovups ymmword ptr [rax], ymm2
vmovups ymmword ptr [rax + rdx - 0x20], ymm0
vzeroupper
ret

; the $128 case jumps to here and does a more stores to finialize the 2 pairs of loads
; that would have happened before the branch to here
.FINAL_128:
vmovups ymmword ptr [rax], ymm2
vmovups ymmword ptr [rax + rdx - 0x20], ymm0
vmovups ymmword ptr [rax + 0x20], ymm3
vmovups ymmword ptr [rax + rdx - 0x40], ymm1
vzeroupper
ret

.FINAL_192:
vmovups ymmword ptr [rax], ymm2
vmovups ymmword ptr [rax + rdx - 0x20], ymm0
vmovups ymmword ptr [rax + 0x20], ymm3
vmovups ymmword ptr [rax + rdx - 0x40], ymm1
vmovups ymmword ptr [rax + 0x40], ymm5
vmovups ymmword ptr [rax + rdx - 0x60], ymm4
vzeroupper
ret
 ```

you can see that LLVM doesn't see that it's generating the same stores 3 times, where it could have generated it a single time and simply jumped to different parts of the fallthrough tree.

In conclusion, the `@expect` built-in was crucial to porting this implementation and provided positive results in relation to the current `@memcpy` implementation.

### CODE REVIEW

**Review by kprotty — APPROVED** [review]

**Comment by dweiller** [comment]
What is the behaviour with `else if`? Should one think of it as an ordered process where you consider the branches in sequence, or would these be equivalent:
```zig
if (val1) {
    // val1 == true code
} else if (@expect(val2, true)) {
    // val2 == true code
} else {
    // ...
}

if (@expect(val2, true)) {
    /// val2 == true code
} else if (val1) {
    // val1 == true code
} else {
    // ...
}
```

**Comment by rohlem** [comment]
@dweiller since `@expect` isn't supposed to change behavior I assume you meant for the second version to be

```zig
if (@expect((!val1) and val2, true)) {
    /// val2 == true code
} else if (val1) {
    // val1 == true code
} else {
    // ...
}
```

**Comment by Rexicon226** [comment]
> What is the behaviour with `else if`? Should one think of it as an ordered process where you consider the branches in sequence, or would these be equivalent?

Semantically nothing changes in the code.

Remember that an `if else` is semantically equivalent to:

```zig
if (val1) {
    // val1 == true code
} else {
  if (@expect(val2, true)) {
    // val2 == true code
  } else {
    // ...
  }
}
```
The `@expect` suggestion applies to a specific `if` statement.

**Comment by dweiller** [comment]
> @dweiller since `@expect` isn't supposed to change behavior I assume you meant for the second version to be
> 
> ```zig
> if (@expect((!val1) and val2, true)) {
>     /// val2 == true code
> } else if (val1) {
>     // val1 == true code
> } else {
>     // ...
> }
> ```

Yea - I messed up those conditionals, they were meant to be mutually exclusive.

```zig
if (val1 and !val2) {
    // val1 == true code
} else if (@expect(val2 and !val1, true)) {
    // val2 == true code
} else {
    // ...
}

if (@expect(val2 and !val1, true)) {
    /// val2 == true code
} else if (val1 and !val2) {
    // val1 == true code
} else {
    // ...
}
```

**Comment by henrikolsen549** [comment]
Note that #489 has a reasonable [proposal](https://github.com/ziglang/zig/issues/489#issuecomment-2099172754) for a `@branchWeight` builtin that I feel hasn't been sufficiently discussed.

**Comment by Inve1951** [comment]
> #### Why it takes a bool
> 
> The `@expect` built-in affects the generation of branching control flow and you can only branch on a bool. To branch on any other value requires an existing comparison. The built-in expect from C, `built_expect`, compares two values producing a bool to the same effect as `@expect(x == y, true)`.

Have you taken branching on optionals into consideration? Maybe those should be supported by this feature. E.g.

```zig
if (opt_val) |val| {
  // unlikely
}
while (it.next()) |val| {
  // unlikely
}
```
Or even
```zig
if (error_union) |val| {
  // unlikely
} else |err| {
  // more likely
}
```

**Comment by andrewrk** [comment]
`@expect` was rejected. Please see #21148 for the alternative accepted proposal. Please note in particular the "issue close criteria".
